vpr/src/place/place.cpp

```diff
@@ -2519,14 +2944,46 @@ static double get_net_wirelength_estimate(ClusterNetId net_id, t_bb* bbptr) {
     /* Cost = wire length along channel * cross_count / average      *
      * channel capacity.   Do this for x, then y direction and add.  */
 
-    ncost = (bbptr->xmax - bbptr->xmin + 1) * crossing;
+    ncost = (bbptr.xmax - bbptr.xmin + 1) * crossing;
+
+    ncost += (bbptr.ymax - bbptr.ymin + 1) * crossing;
+
+    return (ncost);
+}
+
+static double get_net_layer_wirelength_estimate(ClusterNetId /* net_id */,
+                                                const std::vector<t_2D_bb>& bbptr,
+                                                const std::vector<int>& layer_pin_sink_count) {
+    /* WMF: Finds the estimate of wirelength due to one net by looking at   *
+     * its coordinate bounding box.                                         */
+
+    double ncost = 0.;
+    double crossing = 0.;
+    int num_layers = g_vpr_ctx.device().grid.get_num_layers();
+
+    for (int layer_num = 0; layer_num < num_layers; layer_num++) {
+        VTR_ASSERT(layer_pin_sink_count[layer_num] != OPEN);
+        if (layer_pin_sink_count[layer_num] == 0) {
+            continue;
+        }
+        crossing = wirelength_crossing_count(layer_pin_sink_count[layer_num] + 1);
```

Comment:
This may be overkill; if it is making you keep per-layer bb data and sink counts when computing a 3D bounding box, you could simplify it to use the crossing count based on the fanout of the whole net.  It will make little difference since most nets are low fanout, and it isn't clear which approach is more accurate anyway.

Reply:
I am storing the number of sinks that each net has on each layer. The computation doesn't require many operations. For low fanout nets, we already iterate over all sinks, and for high-fanout ones, I update it incrementally.